taf/git.py

```diff
@@ -841,58 +861,89 @@ class NamedGitRepository(GitRepository):
+          name: repository's path relative to the root directory root_dir
+          urls: repository's urls (optional)
+          custom: a dictionary containing other data (optional)
           default_branch: repository's default branch
         path is the absolute path to this repository. It is set by joining
-        root_dir and repo_name.
+        root_dir and name.
         """
         self.root_dir = root_dir
-        self.name = repo_name
-        path = self._get_repo_path(root_dir, repo_name)
+        self.name = name
+        path = self._get_repo_path(root_dir, name)
         super().__init__(
             path,
-            repo_name=repo_name,
-            repo_urls=repo_urls,
-            additional_info=additional_info,
+            name=name,
+            urls=urls,
+            custom=custom,
             default_branch=default_branch,
         )
 
-    def _get_repo_path(self, root_dir, repo_name):
+    @classmethod
+    def from_json_dict(cls, json_data):
+        json_data.pop("path", None)
+        root_dir = json_data.pop("root_dir")
+        urls = json_data.pop("urls")
+        custom = json_data.pop("custom", None)
+        name = json_data.pop("name")
+        default_branch = json_data.pop("default_branch", "master")
+        return cls(root_dir, name, urls, custom, default_branch, **json_data)
+
+    def to_json_dict(self):
+        return {
+            "root_dir": str(self.root_dir),
+            "name": self.name,
+            "path": self.path,
+            "urls": self.urls,
+            "custom": self.custom,
+            "default_branch": self.default_branch,
+        }
+
+    def _get_repo_path(self, root_dir, name):
         """
         get the path to a repo and ensure it is valid.
         (since this is coming from potentially untrusted data)
         """
-        self._validate_repo_name(repo_name)
-        repo_dir = str((Path(root_dir) / (repo_name or "")))
+        self._validate_repo_name(name)
+        repo_dir = str((Path(root_dir) / (name or "")))
         if not repo_dir.startswith(repo_dir):
             self._log_error("repository name is not valid")
-            raise InvalidRepositoryError(f"Invalid repository name: {repo_name}")
+            raise InvalidRepositoryError(f"Invalid repository name: {name}")
         return repo_dir
 
-    def _validate_repo_name(self, repo_name):
+    def _validate_repo_name(self, name):
         """ Ensure the repo name is not malicious """
-        match = _repo_name_re.match(repo_name)
+        match = _repo_name_re.match(name)
         if not match:
             self._log_error("repository name is not valid")
             raise InvalidRepositoryError(
                 "Repository name must be in format namespace/repository "
                 "and can only contain letters, numbers, underscores and "
-                f'dashes, but got "{repo_name}"'
+                f'dashes, but got "{name}"'
             )
 
 
+def get_last_remote_commit(url):
```

Comment:
We use this in our implementation of the publish server and I moved it back when I was trying to move the majority of the code, having misunderstood the idea. TAF does not need it. Not sure if we should keep it here or not. It does not make sense to call this if `url` is `None`, but that has never happened in our case. It does not make much difference for us, though it might make more sense to raise an error if we decide to leave this function.